File: skycli/tests/smoke_tests.py
import grpc
import time
import pytest
import concurrent.futures
import google.protobuf.empty_pb2

from proto.skyvault.v1 import skyvault_pb2
from proto.skyvault.v1 import skyvault_pb2_grpc
from minikube import setup_connection

#
# Fixtures
#


def create_connection(service_name):
    """Creates and manages a connection to a skyvault service via minikube."""
    try:
        channel = setup_connection(f"skyvault-{service_name}")
        print(f"Successfully connected to skyvault-{service_name} service")
        return channel
    except RuntimeError as e:
        pytest.fail(f"Failed to connect to {service_name} gRPC service: {e}")
    except grpc.FutureTimeoutError:
        pytest.fail(f"Failed to connect to {service_name} gRPC service: timeout")


@pytest.fixture(scope="session")
def connections():
    services = ["writer", "reader", "orchestrator"]
    with concurrent.futures.ThreadPoolExecutor(max_workers=len(services)) as executor:
        future_to_service = {
            executor.submit(create_connection, service): service for service in services
        }
        connections = {}
        for future in concurrent.futures.as_completed(future_to_service):
            service = future_to_service[future]
            connections[service] = future.result()
    return connections


@pytest.fixture(scope="session")
def stubs(connections):
    """Creates gRPC stubs for different services."""
    return {
        "writer": skyvault_pb2_grpc.WriterServiceStub(connections["writer"]),
        "reader": skyvault_pb2_grpc.ReaderServiceStub(connections["reader"]),
        "orchestrator": skyvault_pb2_grpc.OrchestratorServiceStub(
            connections["orchestrator"]
        ),
    }


#
# Helper functions
#


def create_table(stub, table_name):
    try:
        request = skyvault_pb2.CreateTableRequest(
            config=skyvault_pb2.TableConfig(table_name=table_name)
        )
        response = stub.CreateTable(request, timeout=10)
        return response.seq_no
    except grpc.RpcError as e:
        if e.code() != grpc.StatusCode.ALREADY_EXISTS:
            raise


def perform_write(stub, table_name, key, value_bytes):
    """Sends a WriteBatch request."""
    write_item = skyvault_pb2.WriteBatchItem(key=key, value=value_bytes)
    table_request = skyvault_pb2.TableWriteBatchRequest(
        table_name=table_name, items=[write_item]
    )
    request = skyvault_pb2.WriteBatchRequest(tables=[table_request])

    start_time = time.time()
    while time.time() - start_time < 5:
        try:
            return stub.WriteBatch(request, timeout=15).seq_no
        except grpc.RpcError as e:
            if e.code() != grpc.StatusCode.NOT_FOUND:
                raise
            print(f"Write failed: {e}")
            time.sleep(1)

    raise grpc.RpcError("Write failed after retries")


def perform_read(stub, table_name, key, expected_value_bytes):
    """Sends a GetBatch request."""
    table_request = skyvault_pb2.TableGetBatchRequest(table_name=table_name, keys=[key])
    request = skyvault_pb2.GetBatchRequest(tables=[table_request])
    response = stub.GetBatch(request, timeout=15)
    if response.tables and response.tables[0].table_name == table_name:
        items = {item.key: item.value for item in response.tables[0].items}
        if key in items and items[key] == expected_value_bytes:
            return True
    return False


def perform_read_with_retry(
    stub, table_name, seq_no, key, expected_value_bytes, retries=30, delay=1
):
    """Attempts to read a key, retrying until the expected value is found or retries run out."""
    table_request = skyvault_pb2.TableGetBatchRequest(table_name=table_name, keys=[key])
    request = skyvault_pb2.GetBatchRequest(tables=[table_request])

    for attempt in range(retries):
        try:
            response = stub.GetBatch(request, timeout=10)
            if response.tables and response.tables[0].table_name == table_name:
                items = {item.key: item.value for item in response.tables[0].items}
                if key in items and items[key] == expected_value_bytes:
                    return True
        except grpc.RpcError as e:
            # Only retry if we get failed_precondition (seq_no mismatch)
            if e.code() != grpc.StatusCode.FAILED_PRECONDITION:
                raise

        time.sleep(delay)

    return False


def trigger_wal_compaction(stub):
    request = skyvault_pb2.KickOffJobRequest(
        params=skyvault_pb2.JobParams(
            wal_compaction=google.protobuf.empty_pb2.Empty(),
        ),
    )
    response = stub.KickOffJob(request, timeout=20)

    # Wait for compaction job to complete with timeout
    start_time = time.time()
    while time.time() - start_time < 5:
        status_request = skyvault_pb2.GetJobStatusRequest(job_id=response.job_id)
        status_response = stub.GetJobStatus(status_request, timeout=10)
        if not status_response.status.pending:
            return status_response.status.seq_no
        time.sleep(0.5)

    raise TimeoutError("Compaction job did not complete within 5 seconds")


def persist_snapshot(stub):
    request = skyvault_pb2.PersistSnapshotRequest()
    response = stub.PersistSnapshot(request, timeout=10)
    return response.seq_no


@pytest.mark.smoke
def test_simple_write_and_read(stubs):
    """Test basic write and read operations work correctly."""
    writer_stub = stubs["writer"]
    reader_stub = stubs["reader"]
    orchestrator_stub = stubs["orchestrator"]

    table_name = "test_simple_write_and_read"
    key = "test_simple_write_and_read_key"
    value = b"test_simple_write_and_read_value"

    create_table(orchestrator_stub, table_name)

    # Write key-value pair
    seq_no = perform_write(writer_stub, table_name, key, value)

    # Verify read works
    assert perform_read_with_retry(reader_stub, table_name, seq_no, key, value), (
        f"Failed to read back key '{key}' after writing"
    )


@pytest.mark.smoke
def test_write_compact_read(stubs):
    """Test that data remains readable after WAL compaction."""
    writer_stub = stubs["writer"]
    reader_stub = stubs["reader"]
    orchestrator_stub = stubs["orchestrator"]

    table_name = "test_write_compact_read"
    key_one = "test_write_compact_read_key_one"
    value_one = b"test_write_compact_read_value_one"
    key_two = "test_write_compact_read_key_two"
    value_two = b"test_write_compact_read_value_two"

    create_table(orchestrator_stub, table_name)

    # Write second key
    seq_no = perform_write(writer_stub, table_name, key_one, value_one)
    seq_no = perform_write(writer_stub, table_name, key_two, value_two)

    # Verify second key is readable before compaction
    assert perform_read_with_retry(
        reader_stub, table_name, seq_no, key_two, value_two
    ), f"Failed to read back key '{key_two}' before compaction"

    # Trigger compaction
    seq_no = trigger_wal_compaction(orchestrator_stub)

    # Verify second key is still readable after compaction
    assert perform_read_with_retry(
        reader_stub, table_name, seq_no, key_two, value_two
    ), f"Failed to read back key '{key_two}' after compaction"

    # Verify first key is also still readable after compaction
    assert perform_read_with_retry(
        reader_stub, table_name, seq_no, key_one, value_one
    ), f"Failed to read back key '{key_one}' after compaction"


@pytest.mark.smoke
def test_snapshot_persistence(stubs):
    """Test that snapshots are persisted correctly."""
    writer_stub = stubs["writer"]
    reader_stub = stubs["reader"]
    orchestrator_stub = stubs["orchestrator"]

    table_name = "test_snapshot_persistence"
    key = "test_snapshot_persistence_key"
    value = b"test_snapshot_persistence_value"

    create_table(orchestrator_stub, table_name)

    # Write some data
    seq_no = perform_write(writer_stub, table_name, key, value)

    # Trigger snapshot persistence
    persist_snapshot(orchestrator_stub)

    # Verify snapshot is persisted
    assert perform_read_with_retry(reader_stub, table_name, seq_no, key, value), (
        f"Failed to read back key '{key}' after snapshot persistence"
    )

File: tests/smoke_tests.rs
use std::time::Duration;

use anyhow::Result;
use futures::TryFutureExt;
use skyvault::proto;
use tokio::process::Child;
use tokio_retry::strategy::FixedInterval;
use tonic::transport::Channel;

mod common;

fn retry_strategy() -> impl Iterator<Item = Duration> {
    FixedInterval::from_millis(100).take(50) // 50 attempts * 100ms = 5 seconds max
}

struct TestStubs {
    writer: proto::writer_service_client::WriterServiceClient<Channel>,
    reader: proto::reader_service_client::ReaderServiceClient<Channel>,
    orchestrator: proto::orchestrator_service_client::OrchestratorServiceClient<Channel>,

    _writer_child: Child,
    _reader_child: Child,
    _orchestrator_child: Child,
}

async fn setup_stubs() -> Result<TestStubs> {
    let ((writer_channel, writer_child), (reader_channel, reader_child), (orchestrator_channel, orchestrator_child)) =
        tokio::try_join!(
            common::setup_connection("skyvault-writer")
                .map_err(|e| anyhow::anyhow!("Failed to setup writer connection: {e}")),
            common::setup_connection("skyvault-reader")
                .map_err(|e| anyhow::anyhow!("Failed to setup reader connection: {e}")),
            common::setup_connection("skyvault-orchestrator")
                .map_err(|e| anyhow::anyhow!("Failed to setup orchestrator connection: {e}"))
        )?;

    Ok(TestStubs {
        writer: proto::writer_service_client::WriterServiceClient::new(writer_channel),
        reader: proto::reader_service_client::ReaderServiceClient::new(reader_channel),
        orchestrator: proto::orchestrator_service_client::OrchestratorServiceClient::new(orchestrator_channel),

        _writer_child: writer_child,
        _reader_child: reader_child,
        _orchestrator_child: orchestrator_child,
    })
}

async fn create_table(
    orchestrator: &mut proto::orchestrator_service_client::OrchestratorServiceClient<Channel>,
    table_name: &str,
) -> Result<Option<i64>> {
    let request = proto::CreateTableRequest {
        config: Some(proto::TableConfig {
            table_id: 0, // This will be assigned by the server
            table_name: table_name.to_string(),
        }),
    };

    match orchestrator.create_table(request).await {
        Ok(response) => Ok(Some(response.into_inner().seq_no)),
        Err(e) if e.code() == tonic::Code::AlreadyExists => Ok(None),
        Err(e) => Err(e.into()),
    }
}

async fn perform_write(
    writer: &mut proto::writer_service_client::WriterServiceClient<Channel>,
    table_name: &str,
    key: &str,
    value_bytes: Vec<u8>,
) -> Result<i64> {
    let write_item = proto::WriteBatchItem {
        key: key.to_string(),
        operation: Some(proto::write_batch_item::Operation::Value(value_bytes)),
    };

    let table_request = proto::TableWriteBatchRequest {
        table_name: table_name.to_string(),
        items: vec![write_item],
    };

    let request = proto::WriteBatchRequest {
        tables: vec![table_request],
    };

    for delay in retry_strategy() {
        match writer.write_batch(request.clone()).await {
            Ok(response) => return Ok(response.into_inner().seq_no),
            Err(e) if e.code() == tonic::Code::NotFound => {
                tokio::time::sleep(delay).await;
                continue;
            },
            Err(e) => return Err(e.into()),
        }
    }

    Err(anyhow::anyhow!("Write failed after retries"))
}

async fn perform_read_with_retry(
    reader: &mut proto::reader_service_client::ReaderServiceClient<Channel>,
    table_name: &str,
    seq_no: i64,
    key: &str,
    expected_value_bytes: &[u8],
) -> Result<bool> {
    let table_request = proto::TableGetBatchRequest {
        table_name: table_name.to_string(),
        keys: vec![key.to_string()],
    };

    let request = proto::GetBatchRequest {
        seq_no,
        tables: vec![table_request],
    };

    for delay in retry_strategy() {
        match reader.get_batch(request.clone()).await {
            Ok(response) => {
                let response_inner = response.into_inner();
                if let Some(table_response) = response_inner.tables.first() {
                    if table_response.table_name == table_name {
                        for item in &table_response.items {
                            if item.key == key && item.value == expected_value_bytes {
                                return Ok(true);
                            }
                        }
                    }
                }

                tokio::time::sleep(delay).await;
            },
            Err(e) if e.code() == tonic::Code::FailedPrecondition => {
                tokio::time::sleep(delay).await;
            },
            Err(e) => return Err(e.into()),
        }
    }

    Ok(false)
}

async fn trigger_wal_compaction(
    orchestrator: &mut proto::orchestrator_service_client::OrchestratorServiceClient<Channel>,
) -> Result<i64> {
    let request = proto::KickOffJobRequest {
        params: Some(proto::JobParams {
            params: Some(proto::job_params::Params::WalCompaction(())),
        }),
    };

    let response = orchestrator.kick_off_job(request).await?;
    let job_id = response.into_inner().job_id;

    // Wait for compaction job to complete with retry
    for delay in retry_strategy() {
        let status_request = proto::GetJobStatusRequest { job_id };
        let status_response = orchestrator.get_job_status(status_request).await?;
        let status = status_response.into_inner().status;

        if let Some(status) = status {
            match status.status {
                Some(proto::job_status::Status::Pending(_)) => {
                    tokio::time::sleep(delay).await;
                    continue;
                },
                Some(proto::job_status::Status::SeqNo(seq_no)) => {
                    return Ok(seq_no);
                },
                Some(proto::job_status::Status::Failed(_)) => {
                    return Err(anyhow::anyhow!("Compaction job failed"));
                },
                None => {
                    return Err(anyhow::anyhow!("Invalid job status"));
                },
            }
        }
    }

    Err(anyhow::anyhow!("Compaction job did not complete within timeout"))
}

async fn persist_snapshot(
    orchestrator: &mut proto::orchestrator_service_client::OrchestratorServiceClient<Channel>,
) -> Result<i64> {
    let request = proto::PersistSnapshotRequest {};
    let response = orchestrator.persist_snapshot(request).await?;
    Ok(response.into_inner().seq_no)
}

#[tokio::test]
#[ignore]
async fn test_simple_write_and_read() -> Result<()> {
    let mut stubs = setup_stubs().await?;

    let table_name = "test_simple_write_and_read";
    let key = "test_simple_write_and_read_key";
    let value = b"test_simple_write_and_read_value".to_vec();

    create_table(&mut stubs.orchestrator, table_name).await?;

    // Write key-value pair
    let seq_no = perform_write(&mut stubs.writer, table_name, key, value.clone()).await?;

    // Verify read works
    let read_success = perform_read_with_retry(&mut stubs.reader, table_name, seq_no, key, &value).await?;

    assert!(read_success, "Failed to read back key '{key}' after writing");

    Ok(())
}

#[tokio::test]
#[ignore]
async fn test_write_compact_read() -> Result<()> {
    let mut stubs = setup_stubs().await?;

    let table_name = "test_write_compact_read";
    let key_one = "test_write_compact_read_key_one";
    let value_one = b"test_write_compact_read_value_one".to_vec();
    let key_two = "test_write_compact_read_key_two";
    let value_two = b"test_write_compact_read_value_two".to_vec();

    create_table(&mut stubs.orchestrator, table_name).await?;

    // Write both keys
    let _seq_no = perform_write(&mut stubs.writer, table_name, key_one, value_one.clone()).await?;
    let seq_no = perform_write(&mut stubs.writer, table_name, key_two, value_two.clone()).await?;

    // Verify second key is readable before compaction
    let read_success = perform_read_with_retry(&mut stubs.reader, table_name, seq_no, key_two, &value_two).await?;

    assert!(read_success, "Failed to read back key '{key_two}' before compaction");

    // Trigger compaction
    let seq_no = trigger_wal_compaction(&mut stubs.orchestrator).await?;

    // Verify second key is still readable after compaction
    let read_success = perform_read_with_retry(&mut stubs.reader, table_name, seq_no, key_two, &value_two).await?;

    assert!(read_success, "Failed to read back key '{key_two}' after compaction");

    // Verify first key is also still readable after compaction
    let read_success = perform_read_with_retry(&mut stubs.reader, table_name, seq_no, key_one, &value_one).await?;

    assert!(read_success, "Failed to read back key '{key_one}' after compaction");

    Ok(())
}

#[tokio::test]
#[ignore]
async fn test_snapshot_persistence() -> Result<()> {
    let mut stubs = setup_stubs().await?;

    let table_name = "test_snapshot_persistence";
    let key = "test_snapshot_persistence_key";
    let value = b"test_snapshot_persistence_value".to_vec();

    create_table(&mut stubs.orchestrator, table_name).await?;

    // Write some data
    let seq_no = perform_write(&mut stubs.writer, table_name, key, value.clone()).await?;

    // Trigger snapshot persistence
    persist_snapshot(&mut stubs.orchestrator).await?;

    // Verify snapshot is persisted
    let read_success = perform_read_with_retry(&mut stubs.reader, table_name, seq_no, key, &value).await?;

    assert!(
        read_success,
        "Failed to read back key '{key}' after snapshot persistence"
    );

    Ok(())
}
